src/backend/opencl/kernel/reduce.hpp

```diff
@@ -295,15 +295,15 @@ To reduce_all(Param in, int change_nan, double nanval) {
                                      sizeof(Ti) * in.info.offset,
                                      sizeof(Ti) * in_elements, h_ptr.data());
 
-        Transform<Ti, To, op> transform;
-        Binary<To, op> reduce;
-        compute_t<To> out       = Binary<To, op>::init();
-        compute_t<To> nanval_to = scalar<To>(nanval);
+        Transform<Ti, compute_t<To>, op> transform;
+        Binary<compute_t<To>, op> reduce;
+        compute_t<To> out       = Binary<compute_t<To>, op>::init();
+        compute_t<To> nanval_to = scalar<compute_t<To>>(nanval);
 
         for (int i = 0; i < (int)in_elements; i++) {
-            To in_val = transform(h_ptr[i]);
+            compute_t<To> in_val = transform(h_ptr[i]);
             if (change_nan) in_val = IS_NAN(in_val) ? nanval_to : in_val;
-            out = reduce(out, in_val);
+            out = reduce(out, compute_t<To>(in_val));
```

Comment:
I do see that sort_index is used for implementing topk in OpenCL backend, so changes to that makes sense.

However, I don't see how these are related to this PR ? Are they a fix and accidentally clubbed together ? We need to have a unit test if it is a bug in reduce unit tests ?

Reply:
If you want to compile the topk changes then these changes are also required. I am not able to compile without it.